File: implementations/javascript/src/conversion.ts
/**
 * LSF to JSON conversion utilities
 * 
 * This module provides functions to convert between LSF and JSON formats.
 */

import { LSFDecoder } from './decoder';

/**
 * Convert an LSF string to a JSON string
 * 
 * @param lsfString The LSF formatted string
 * @param space Indentation spacing for pretty-printing (number of spaces or null for compact)
 * @param replacer A function that transforms the results or an array of strings and numbers that serves as an allowlist for selecting properties
 * @returns JSON formatted string
 * 
 * @example
 * ```ts
 * lsfToJson("$o~user$r~$f~name$f~John$r~$f~age$f~30$r~")
 * // Returns: '{"user":{"name":"John","age":"30"}}'
 * 
 * lsfToJson("$o~user$r~$f~name$f~John$r~", 2)
 * // Returns pretty-printed JSON with 2-space indentation
 * ```
 */
export function lsfToJson(
    lsfString: string, 
    space?: number | string,
    replacer?: (this: any, key: string, value: any) => any | Array<number | string>
): string {
    // First decode LSF to JavaScript objects
    const decoder = new LSFDecoder();
    const data = decoder.decode(lsfString);
    
    // Convert to JSON
    return JSON.stringify(data, replacer, space);
}

/**
 * Convert an LSF string to a pretty-printed JSON string
 * 
 * This is a convenience wrapper around lsfToJson with 2-space indentation
 * 
 * @param lsfString The LSF formatted string
 * @returns Pretty-printed JSON string
 * 
 * @example
 * ```ts
 * lsfToJsonPretty("$o~user$r~$f~name$f~John$r~")
 * // Returns: pretty-printed JSON with 2-space indentation
 * ```
 */
export function lsfToJsonPretty(lsfString: string): string {
    return lsfToJson(lsfString, 2);
} 

File: implementations/python/lsf/conversion.py
"""
LSF to JSON conversion utilities

This module provides functions to convert between LSF and JSON formats.
"""

import json
from typing import Any, Dict, Optional, Union

from .decoder import LSFDecoder


def lsf_to_json(
    lsf_string: str, 
    indent: Optional[int] = None, 
    sort_keys: bool = False
) -> str:
    """
    Convert an LSF string to a JSON string
    
    Args:
        lsf_string: The LSF formatted string
        indent: Indentation level for pretty-printing (None for compact)
        sort_keys: Whether to sort dictionary keys in the output
        
    Returns:
        JSON formatted string
        
    Example:
        >>> lsf_to_json("$o~user$r~$f~name$f~John$r~$f~age$f~30$r~")
        '{"user":{"name":"John","age":"30"}}'
        
        >>> lsf_to_json("$o~user$r~$f~name$f~John$r~", indent=2)
        '{
          "user": {
            "name": "John"
          }
        }'
    """
    # First decode LSF to Python objects
    decoder = LSFDecoder()
    data = decoder.decode(lsf_string)
    
    # Convert to JSON
    return json.dumps(data, indent=indent, sort_keys=sort_keys)


def lsf_to_json_pretty(lsf_string: str) -> str:
    """
    Convert an LSF string to a pretty-printed JSON string
    
    This is a convenience wrapper around lsf_to_json with indent=2
    
    Args:
        lsf_string: The LSF formatted string
        
    Returns:
        Pretty-printed JSON string
        
    Example:
        >>> lsf_to_json_pretty("$o~user$r~$f~name$f~John$r~")
        '{
          "user": {
            "name": "John"
          }
        }'
    """
    return lsf_to_json(lsf_string, indent=2) 